### PR TITLE
fix "no compatible architectures" in riscv32

### PR DESCRIPTION
**Pull request by U2FsdGVkX1**
https://github.com/rpm-software-management/rpm/blob/master/lib/rpmrc.c#L1334
because in defaultMachine, riscv32 will be assigned as "riscv32", however rpmrc identifies rv32 as "riscv"
this will cause build some packages in riscv32 to say "no compatible architectures found for build"

### CODE REVIEW

**Comment by pmatilai** [comment]
This looks risky as it effectively renames an architecture, plus plain wrong in places (more on those shortly). But, I can see the point: plain "riscv" is not even supposed to exist, it's always translated to riscv\<bitness\>.

@rwmjones , thoughts?

File: rpmrc.in [hunk 1]
```diff
@@ -404,7 +405,7 @@ buildarchtranslate: sh4a: sh4
 
 buildarchtranslate: aarch64: aarch64
 
-buildarchtranslate: riscv: riscv64
+buildarchtranslate: riscv32: riscv64
```
Comment (pmatilai) [inline rpmrc.in:408]: This seems wrong. Surely you don't want riscv32 to generate riscv64 packages by default?

Comment (U2FsdGVkX1) [reply rpmrc.in:408]: > This seems wrong. Surely you don't want riscv32 to generate riscv64 packages by default?

yes this should be fixed!

File: rpmrc.in [hunk 2]
```diff
@@ -268,7 +269,7 @@ arch_canon:	mipsr6el: mipsr6el	20
 arch_canon:	mips64r6: mips64r6	21
 arch_canon:	mips64r6el: mips64r6el	21
 
-arch_canon:	riscv: riscv64	22
+arch_canon:	riscv32: riscv64	22
```
Comment (pmatilai) [inline rpmrc.in:272]: This doesn't look right. Should probably be "riscv32: riscv32 22" instead.

**Comment by rwmjones** [comment]
I asked @davidlt to take a look, but just a few things ...

We don't use 32 bit RISC-V in Fedora, and don't plan to, so to some extent we don't care about it.  Note that 32 bit RISC-V machines are currently confined to tiny embedded systems that couldn't possibly run Linux.  eg. They lack userspace mode, MMU, enough RAM, etc.  Maybe they could in future, although it seems unlikely to ever be interesting to us.

We don't use the base `riscv` architecture either.  Our RPMs are `riscv64`.

**Comment by davidlt** [comment]
To my knowledge none of main distributions support 32-bit RISCV. OpenEmbedde definitely has it. Most likely the same applies to Buildroot too. There are no native riscv32 hardware available that would target Linux to my knowledge. It's mainly tested with QEMU.

IIRC Alibaba/T-HEAD cores are `CONFIG_COMPAT` capable. At least of Fedora/RISCV side we have no plans to enable CONFIG_COMPAT.

TL;DR There is very little interest in 32-bit RISCV + feature rich operating systems. Even the tiny and cheap boards with (or without) MMU that are Linux capable are 64-bit today.

I don't think that anything uses `riscv` arch string. I didn't know it even existed. Everyone expects `riscv32`, `riscv64` and there are several derivates in other places (e.g. OpenEmbedded, FreeBSD IIRC).

**Comment by pmatilai** [comment]
Thanks @rwmjones and @davidlt for the insight, I had no clue.

@U2FsdGVkX1 - what's the environment you're running into this? If there's an actual use-case (distro or such) for riscv-32 now, it wouldn't hurt to have people aware of it.

**Comment by U2FsdGVkX1** [comment]
@pmatilai Hi, I'm currently a plct's internship and working on the Fedora RISCV 32 bit bootstrap
due to this, I need to compile many rpm packages and noarch architectures will not compile in this environment, it says "no compatible architectures found for build"
so I've simply fixed the problem, and maybe there's something else I haven't fixed

**Comment by pmatilai** [comment]
Okay, Fedora Riscv-32bit bootstrap. It seems a bit odd considering the comments above, but that's not a blocker of course.

These will need more elaborate commit messages to summarize why this apparent breakage is not (summarize the above discussion). Also you'll need to supply a real name in the git commit, we don't accept patches from pseudonyms.

**Comment by Conan-Kudo** [comment]
Yocto will probably wind up using this too, since they support the RPM stack with architectures like this.

**Comment by ffesti** [comment]
OK, still someone needs to squash this and provide a proper commit message. I'll close this soonish otherwise.
